rundeckapp/grails-app/utils/rundeck/codecs/JobsXMLCodec.groovy

```diff
@@ -282,42 +313,70 @@ class JobsXMLCodec {
-                            throw new JobXMLException("${trigger} plugin had blank or missing 'type' attribute")
-                        }
-                        if(map.notification[trigger].plugin instanceof Collection && !map.notification[trigger].plugin.every{it.type}){
-                            throw new JobXMLException("${trigger} plugin had blank or missing 'type' attribute")
-                        }
+                    if (notifData.type==null) {
+                        throw new JobXMLException("${trigger} plugin had blank or missing 'type' attribute")
                     }
-                    if(map.notification[trigger].webhook){
-                        map.notification[trigger].urls = map.notification[trigger].webhook.remove('urls')
-                        map.notification[trigger].remove('webhook')
+                    if (notifData.configuration==null) {
+                        throw new JobXMLException("${trigger} plugin had blank or missing 'configuration' element")
                     }
-                    if(map.notification[trigger].plugin && map.notification[trigger].plugin instanceof Map){
-                        convertPluginToMap(map.notification[trigger].plugin)
-                    }else if(map.notification[trigger].plugin && map.notification[trigger].plugin instanceof Collection){
-                        map.notification[trigger].plugin.each{
-                            convertPluginToMap(it)
-                        }
+                    convertPluginToMap(notifData)
+                    Map plugNotifs = map.notification[trigger].find {notif -> notif.plugin}
+                    if(!plugNotifs)  map.notification[trigger].add(['plugin':[notifData]])
+                    else             plugNotifs.plugin.add(notifData)
+                }
+            }
+
+            boolean isOldFormat = useOldFormat(map.notification, false)
+            triggers.each{trigger->
+                if((map.notification[trigger].size() < 1) || (map.notification[trigger].email==null && map.notification[trigger].webhook==null && map.notification[trigger].plugin==null)){
+                    throw new JobXMLException("notification '${trigger}' element had missing 'email' or 'webhook' or 'plugin' element")
+                }
+
+                final def xmlNotif = map.notification[trigger]
+                map.notification[trigger] = []
+                if(isOldFormat && xmlNotif['webhook']){
+                    if(xmlNotif['format'] && xmlNotif['httpMethod']){
+                        xmlNotif['webhook']['format'] = xmlNotif.remove('format')
+                        xmlNotif['webhook']['httpMethod'] = xmlNotif.remove('httpMethod')
                     }
                 }
+
+                xmlNotif.each { notifEntry ->
+                        final String notifType = notifEntry.getKey()
+                        if(notifType.equals("email") || notifType.equals("webhook") || notifType.equals("plugin")) {
```

Comment:
are these the only notification types supported?

Reply:
Yes, in the future if we need to add more we could use the plugin notification map structure. Actually, at some point we should end up using same notification map structure for all of them